x/outline-fetch/main.go

```diff
@@ -0,0 +1,131 @@
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/transport/shadowsocks"
+	"github.com/Jigsaw-Code/outline-sdk/transport/socks5"
+	"github.com/Jigsaw-Code/outline-sdk/transport/split"
+)
+
+func makeStreamDialer(transportConfig string) (transport.StreamDialer, error) {
+	if transportConfig == "" {
+		return &transport.TCPStreamDialer{}, nil
+	}
+
+	accessKeyURL, err := url.Parse(transportConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse access key: %w", err)
+	}
+	switch accessKeyURL.Scheme {
+
+	case "socks5":
+		return socks5.NewStreamDialer(&transport.TCPEndpoint{Address: accessKeyURL.Host})
+
+	case "ss":
+		cipherInfoBytes, err := base64.URLEncoding.WithPadding(base64.NoPadding).DecodeString(accessKeyURL.User.String())
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode cipher info [%v]: %v", accessKeyURL.User.String(), err)
+		}
+		cipherName, secret, found := strings.Cut(string(cipherInfoBytes), ":")
+		if !found {
+			return nil, fmt.Errorf("invalid cipher info: no ':' separator")
+		}
+		cryptoKey, err := shadowsocks.NewEncryptionKey(cipherName, secret)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create cipher: %w", err)
+		}
+
+		dialer, err := shadowsocks.NewStreamDialer(&transport.TCPEndpoint{Address: accessKeyURL.Host}, cryptoKey)
+		if err != nil {
+			return nil, err
+		}
+
+		prefixStr := accessKeyURL.Query().Get("prefix")
+		if len(prefixStr) > 0 {
+			prefix, err := parseStringPrefix(prefixStr)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse prefix: %w", err)
+			}
+			dialer.SaltGenerator = shadowsocks.NewPrefixSaltGenerator(prefix)
+		}
+		return dialer, nil
+
+	case "split":
+		splitPoint, err := strconv.Atoi(accessKeyURL.Host)
+		if err != nil {
+			return nil, fmt.Errorf("splitPoint is not a number: %v. Split config should be in split:<number> format", accessKeyURL.Host)
+		}
+		return split.NewStreamDialer(&transport.TCPStreamDialer{}, int64(splitPoint))
+
+	default:
+		return nil, fmt.Errorf("access key scheme %v:// is not supported", accessKeyURL.Scheme)
+	}
+}
+
+func parseStringPrefix(utf8Str string) ([]byte, error) {
+	runes := []rune(utf8Str)
+	rawBytes := make([]byte, len(runes))
+	for i, r := range runes {
+		if (r & 0xFF) != r {
+			return nil, fmt.Errorf("character out of range: %d", r)
+		}
+		rawBytes[i] = byte(r)
+	}
+	return rawBytes, nil
+}
+
+func main() {
+	transportFlag := flag.String("transport", "", "Transport config")
+	flag.Parse()
+
+	url := flag.Arg(0)
+	if url == "" {
+		log.Fatal("Need to pass the URL to fetch in the command-line")
+	}
+
+	dialer, err := makeStreamDialer(*transportFlag)
+	if err != nil {
+		log.Fatalf("Could not create dialer: %v", err)
+	}
+	httpClient := &http.Client{Transport: &http.Transport{DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
+		return dialer.Dial(ctx, addr)
```

Comment:
We might need to make sure `network == "tcp"` here.

Reply:
Done